refactor(core): generate initial inputs from inputs map

crisbeto opened the pull request:
We have some logic that generates the `InitialInputs` data structure by taking the static attributes of a node and looking them up in the `DirectiveDef.inputs` map to determine if they correspond to an input.

This works fine at the moment, but it will make it trickier to generate the correct `InitialInputs` for host directives, because the `DirectiveDef.inputs` entries might be aliased under a different name.

These changes rework the existing `generateInitialInputs` function so that it does its lookup on the remapped inputs that have accounted for the aliasing process already.

crisbeto commented:
Since the inputs change was slightly risky, [I ran a TGP](http://test/OCL:469400652:BASE:469417392:1661251918281:8a1e2ce) to make sure that everything still works. The handful of broken targets are pre-existing breakages.

pkozlowski-opensource reviewed (commented):
Looking briefly at the changes I'm not sure I understand the motivation for the second commit. More specifically, reading through the commit message:

> Currently `resolveDirectives` can be split up into two parts: directive matching (implemented in `findDirectiveDefMatches`) and directive application (implemented inline inside `resolveDirectives`). This will be problematic for host directives, because the logic from `findDirectiveDefMatches` will have to return two pieces of information instead of one: the list of matched directives and the host directive input/output mapping. We could make it work with the current code structure, but we'd have to return a tuple type from `findDirectiveDefMatches` which will consume more memory.

I don't think that `findDirectiveDefMatches` should be responsible for _both_ matching initial directives _and_ resolving host directives. I didn't think about it enough, but to me it seems like those are 2 separate steps. In this sense I would like to preserve a function similar to `findDirectiveDefMatches` - it is a well encapsulated logic that reflects what is happening on a TNode.

As for the "directive application" part - maybe this is just about the name, but I don't feel like "directive application" reflects what is happening. It seems to me like we are doing 2 steps:
* finding matching directives 
* building internal data structures based on the directives found - I guess this is what you mean by "directive application"?

Note that I'm not saying that the current code is well structured - far from it, and we could do several improvements in the area. I'm just not sure if those refactors are moving us in the right direction. 

How about we pair on those bits sometime tomorrow?

crisbeto commented:
@pkozlowski-opensource I've dropped the second commit and kept the changes for generating the initial inputs. Can you take another look?

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1511,7 +1512,15 @@ function generateInitialInputs(inputs: {[key: string]: string}, attrs: TAttribut
-      inputsToStore.push(attrName as string, inputs[attrName as string], attrs[i + 1] as string);
+
+      const inputConfig = inputs[attrName as string];
+      for (let j = 1; j < inputConfig.length; j += 2) {
```
nit: I'd like to propose adding a quick comment here describing why we do additional filtering. The same/similar note can be added to the description of this function for future context.

AndrewKushnir replied:
one more nit: I've spent some time to understand why we start iterating with 1. I think starting from 0 might be a bit more readable :)

AndrewKushnir reviewed (approved):
Reviewed-for: size-tracking

dylhunn reviewed (approved):
reviewed-for: size-tracking

crisbeto committed:
refactor(core): generate initial inputs from inputs map

We have some logic that generates the `InitialInputs` data structure by taking the static attributes of a node and looking them up in the `DirectiveDef.inputs` map to determine if they correspond to an input.

This works fine at the moment, but it will make it trickier to generate the correct `InitialInputs` for host directives, because the `DirectiveDef.inputs` entries might be aliased under a different name.

These changes rework the existing `generateInitialInputs` function so that it does its lookup on the remapped inputs that have accounted for the aliasing process already.

dylhunn commented:
This PR was merged into the repository by commit deaa05a393517fcf9182559d6e637c5929fb814f.
